### DISCUSSION TITLE
Support for `re-install` update strategy

### DISCUSSION THREAD

**Discussion by ennekein** [helmfile/helmfile#1717] [Ideas]
We have situations where we need to change immutable fields in Deployments and StatefulSets when updating releases through helmfile. In these cases, `helmfile apply` will run `helm upgrade` on the said releases and it will fail because of the immutable fields. Using the `--force` option leads to the same result.

Would it be possible to implement a release field that would change the update behavior to subsequent `helm uninstall` and `helm install` calls? Something like `updateStrategy: reinstall`?

I volunteer to implement it if this is something that is desired.

**Comment by ennekein** [comment]
@yxxhero do you have any thoughts on this? Thanks in advance!

**Reply by yxxhero** [reply]
Helmfile follows a declarative specification; re-install seems to contradict this.  
/cc @mumoshu

**Reply by acastong** [reply]
A declarative specification is about declaring the desired end state.  In this case when a helm chart creates a resource that has immutable fields, the only way to get to that desired end state is through deleting and recreating these resource, as unfortunate as that may be.

It would really be great if it was possible to have helmfile only perform this re-install operation if it is truly needed: when helm diff finds that there are changes to apply.

**Comment by acastong** [comment]
@yxxhero and @mumoshu, would you have alternate suggestions on how helmfile could handle the update of releases that include resources with immutable fields?

**Comment by acastong** [comment]
The fact that --force does not work on immutable resources (for helm 3) is documented in issue https://github.com/helm/helm/issues/7082. Within that issue, it was discussed that updating only the resources that cause conflict due to immutable fields could cause problem if only some resources of a helm release are recreated. As such, there has been no driver to handle the situation within helm.

Having the possibility to have helmfile recreate an entire release when a change is detected could prevent these concerns.
